server/resp/src/test/java/org/infinispan/server/resp/TransactionOperationsTest.java

```diff
@@ -94,4 +100,83 @@ public void testStartNestedTx() throws Exception {
 
       multi.closeAsync().get(10, TimeUnit.SECONDS);
    }
+
+   @Test
+   public void testTransactionAbortWithError() {
+      RedisCommands<String, String> redis = redisConnection.sync();
+
+      assertThat(redis.multi()).isEqualTo("OK");
+      assertThat(redisConnection.isMulti()).isTrue();
+
+      // This returns an -ERR, but lettuce just returns null when in TX context.
+      assertThat(redis.watch("something")).isNull();
+
+
+      assertThatThrownBy(redis::exec)
+            .isInstanceOf(RedisCommandExecutionException.class)
+            .hasMessage("EXECABORT Transaction discarded because of previous errors.");
+   }
+
+   @Test
+   public void testTransactionWithWatcher() {
+      RedisCommands<String, String> redis = redisConnection.sync();
+
+      assertThat(redis.watch("tx-watcher-k1")).isEqualTo("OK");
+      assertThat(redis.multi()).isEqualTo("OK");
+      assertThat(redisConnection.isMulti()).isTrue();
+
+      assertThat(redis.set("tx-watcher-k1", "value")).isNull();
+
+      TransactionResult result = redis.exec();
+      assertThat(result.<String>get(0)).isEqualTo("OK");
+
+      assertThat(redisConnection.isMulti()).isFalse();
+
+      assertThat(redis.get("tx-watcher-k1")).isEqualTo("value");
+   }
+
+   @Test
+   public void testWatcherCapturesChange() {
+      testMultiWithWatcher(false);
+   }
+
+   @Test
+   public void testRemovingWatcherBeforeExec() {
+      testMultiWithWatcher(true);
+   }
+
+   private void testMultiWithWatcher(boolean unwatchBeforeExec) {
+      StatefulRedisConnection<String, String> multi = client.connect();
```

Comment:
would be nice to have some comments on which are the commands that are being executed if we try the cli. just as a helper. not asking for that right now, but I found that useful when I started doing it in the sorted sets suite

Reply:
Extended a bit.